src/cargo/core/compiler/custom_build.rs

```diff
@@ -588,7 +588,15 @@ impl BuildOutput {
                         // to set RUSTC_BOOTSTRAP.
                         // If this is a nightly build, setting RUSTC_BOOTSTRAP wouldn't affect the
                         // behavior, so still only give a warning.
-                        if nightly_features_allowed {
+                        // NOTE: cargo only allows nightly features on RUSTC_BOOTSTRAP=1, but we
+                        // want setting any value of RUSTC_BOOTSTRAP to downgrade this to a warning
+                        // (so that `RUSTC_BOOTSTRAP=pkg_name` will work)
+                        let rustc_bootstrap_allows = |name: &str| {
+                            std::env::var("RUSTC_BOOTSTRAP").map_or(false, |var| {
+                                var.split(',').any(|s| s == name)
+                            })
```

Comment:
This is copied from https://github.com/jyn514/rust/blob/0679a4cd93db7dfd33b45cd5bd6be52a0f5e894a/compiler/rustc_feature/src/lib.rs#L86-L89.

Reply:
Hmm, `pkg_name` makes me worried it will have dashes instead of underscores, let me test that.

Reply:
Yup, it does indeed have dashes instead of underscores. @ehuss do you know how to get the --crate-name cargo will pass to rustc?

Reply:
Looks like `unit.target.crate_name()` works.

Reply:
Looks like `unit.target.crate_name()` works.

Reply:
Hmm no, that's the name of the build script, which is always build_script_build. I guess this should be different depending on what targets cargo plans to build? :confused: and there could be multiple since you can e.g. build all binaries at once.

Reply:
Yea, this is maybe going to be a little more complicated.

For the common case, it is probably the library name that it wants to check. That can be obtained from unit with something like `unit.pkg.targets().iter().find(|t| t.is_lib()).map(|t| t.crate_name())`.  

I guess that brings up the question on what to do for binaries, tests, examples, benches, etc.  I'm uncertain how that should be handled, though.  Maybe @joshtriplett has ideas?

Reply:
I would also be ok with reverting https://github.com/rust-lang/cargo/pull/9181 on beta until this is fixed.

Reply:
For now, maybe we should downgrade it to an error if library name matches *or* if there's no library target? I think erring on the side of less hard errors is better if we're not sure, it will still give an unsilencable warning.

Reply:
Is there a way to find reverse dependencies that are currently being compiled? Like if someone passes `cargo build --bin xyz` and that runs a build script, can I get back `xyz` somehow?

Reply:
Another alternative is to give a hard error here - since there's no library, it has no reverse dependencies. I guess that could break people running `cargo install` though?